File: frontend/js/views/pmi_removal.js
// PMI removal timeline — months until the loan amortizes to 80% / 78% of the
// original home value, via /calc/pmi-removal. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const pct = (n) => Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%';
const VIEW = 'pmi-removal';
let lastReport = null;
let lastBody = null;
const months = (n) => {
    if (n == null) return '—';
    const m = Math.round(n);
    const y = Math.floor(m / 12);
    const rem = m % 12;
    return `${m} (${y}y ${rem}m)`;
};

export async function renderPmiRemoval(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.pmi.h1.title">// PMI REMOVAL</span></h1>
        <p class="muted small" data-i18n="view.pmi.hint.intro">
            When private mortgage insurance drops off a conventional loan. PMI is tied to
            loan-to-value against the original home value — you can request cancellation at 80% LTV,
            and the servicer must cancel automatically at 78% (Homeowners Protection Act). This shows
            the month scheduled payments reach each threshold. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.pmi.h2.inputs">The loan</h2>
            <form id="pmi-form" class="inline-form">
                <label><span data-i18n="view.pmi.label.value">Original home value ($)</span>
                    <input type="number" step="0.01" min="0" name="original_home_value_usd" value="400000" required></label>
                <label><span data-i18n="view.pmi.label.loan">Original loan ($)</span>
                    <input type="number" step="0.01" min="0" name="original_loan_usd" value="360000" required></label>
                <label><span data-i18n="view.pmi.label.rate">Annual rate (%)</span>
                    <input type="number" step="0.001" min="0" name="annual_rate_pct" value="6" required></label>
                <label><span data-i18n="view.pmi.label.term">Loan term (months)</span>
                    <input type="number" step="1" min="1" name="loan_term_months" value="360" required></label>
            </form>
            <div id="pmi-tools" class="ce-toolbar"></div>
            <button type="button" id="pmi-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="pmi-sens" class="ce-sens"></div>
        </div>
        <div id="pmi-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#pmi-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        return {
            original_home_value_usd: Number(fd.get('original_home_value_usd')) || 0,
            original_loan_usd: Number(fd.get('original_loan_usd')) || 0,
            annual_rate_pct: Number(fd.get('annual_rate_pct')) || 0,
            loan_term_months: Number(fd.get('loan_term_months')) || 0,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcPmiRemoval(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.pmi.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#pmi-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'pmi-removal.csv' });
    mount.querySelector('#pmi-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['original_ltv_pct', r.original_ltv_pct],
        ['monthly_payment_usd', r.monthly_payment_usd],
        ['months_to_80', r.months_to_80 == null ? '' : r.months_to_80],
        ['months_to_78', r.months_to_78 == null ? '' : r.months_to_78],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#pmi-result');
    // Line chart: months to the 78% auto-cancel threshold as the rate sweeps 3% -> 9%
    // (a higher rate slows principal paydown, pushing PMI removal later).
    const xs = enh.linspace(3, 9, 13);
    const pts = await Promise.all(xs.map(async (rate) => {
        const rr = await api.calcPmiRemoval({ ...body, annual_rate_pct: rate });
        return { x: rate, y: rr && rr.months_to_78 != null ? rr.months_to_78 : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'rate %', ylabel: 'months to 78%' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.pmi.h2.result">When PMI drops</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.pmi.card.ltv">Original LTV</div>
                    <div class="value">${pct(r.original_ltv_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.pmi.card.request">Request at 80%</div>
                    <div class="value">${months(r.months_to_80)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.pmi.card.auto">Auto at 78%</div>
                    <div class="value pos">${months(r.months_to_78)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.pmi.row.payment">Monthly payment</td><td>${money(r.monthly_payment_usd)}</td></tr>
                    <tr><td data-i18n="view.pmi.row.t80">80% balance threshold</td><td>${money(r.target_80_balance_usd)}</td></tr>
                    <tr><td data-i18n="view.pmi.row.t78">78% balance threshold</td><td>${money(r.target_78_balance_usd)}</td></tr>
                    <tr><td data-i18n="view.pmi.row.request">Months to request (80%)</td><td>${months(r.months_to_80)}</td></tr>
                    <tr class="emph"><td data-i18n="view.pmi.row.auto">Months to automatic (78%)</td><td>${months(r.months_to_78)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#pmi-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: annual rate 3% -> 9%; y: original loan 0.8x -> 1.2x. Output: months to 78% (lower better -> negate).
    const loan = base.original_loan_usd || 360000;
    const xVals = enh.linspace(3, 9, 5);
    const yVals = enh.linspace(loan * 0.8, loan * 1.2, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'annual_rate_pct', yKey: 'original_loan_usd', xVals, yVals, compute: (b) => api.calcPmiRemoval(b), pick: (r) => (r && r.months_to_78 != null ? r.months_to_78 : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells: cells.map((row) => row.map((v) => (v == null ? null : -v))), fmt: (v) => (v == null ? '—' : String(-Math.round(v)) + 'mo'), xfmt: (v) => v.toFixed(1) + '%', yfmt: (v) => '$' + Math.round(v / 1000) + 'k', xName: t('view.pmi.label.rate') || 'Rate', yName: t('view.pmi.label.loan') || 'Loan' });
}

File: crates/traderview-core/src/pmi_removal.rs
//! PMI removal timeline — when scheduled amortization drops private mortgage
//! insurance off a conventional loan.
//!
//! PMI is tied to loan-to-value against the *original* home value: a borrower
//! may request cancellation at 80% LTV, and the servicer must cancel
//! automatically at 78% (Homeowners Protection Act). This finds the month the
//! amortizing balance first reaches each threshold.
//!
//! Solving `B_k = P(1+i)^k − pmt·((1+i)^k − 1)/i = target` for k:
//!
//! ```text
//! (1+i)^k = (target − pmt/i) / (P − pmt/i)
//! k = ln(that) / ln(1+i)
//! ```

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct PmiInput {
    pub original_home_value_usd: f64,
    pub original_loan_usd: f64,
    pub annual_rate_pct: f64,
    pub loan_term_months: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct PmiResult {
    pub monthly_payment_usd: f64,
    /// Loan-to-value at origination, percent.
    pub original_ltv_pct: f64,
    /// Balance threshold for borrower-requested cancellation (80% of value).
    pub target_80_balance_usd: f64,
    /// Balance threshold for automatic termination (78% of value).
    pub target_78_balance_usd: f64,
    /// Months until the balance reaches 80% LTV (request); `None` if no PMI.
    pub months_to_80: Option<f64>,
    /// Months until the balance reaches 78% LTV (automatic); `None` if no PMI.
    pub months_to_78: Option<f64>,
}

fn payment(p: f64, i: f64, n: f64) -> f64 {
    if p <= 0.0 || n <= 0.0 {
        return 0.0;
    }
    if i.abs() < 1e-12 {
        p / n
    } else {
        let f = (1.0 + i).powf(n);
        p * i * f / (f - 1.0)
    }
}

/// Whole months until the amortizing balance first reaches `target`.
fn months_until(p: f64, i: f64, pmt: f64, target: f64) -> Option<f64> {
    if p <= target {
        return Some(0.0); // already at or below the threshold
    }
    if i.abs() < 1e-12 {
        // Linear paydown: principal portion is pmt each month.
        if pmt <= 0.0 {
            return None;
        }
        return Some(((p - target) / pmt).ceil());
    }
    let denom = p - pmt / i;
    if denom == 0.0 {
        return None;
    }
    let x = (target - pmt / i) / denom;
    if x <= 0.0 {
        return None;
    }
    Some((x.ln() / (1.0 + i).ln()).ceil())
}

pub fn analyze(input: &PmiInput) -> PmiResult {
    let i = input.annual_rate_pct / 100.0 / 12.0;
    let pmt = payment(input.original_loan_usd, i, input.loan_term_months);

    let t80 = 0.80 * input.original_home_value_usd;
    let t78 = 0.78 * input.original_home_value_usd;

    let ltv = if input.original_home_value_usd > 0.0 {
        input.original_loan_usd / input.original_home_value_usd * 100.0
    } else {
        0.0
    };

    PmiResult {
        monthly_payment_usd: pmt,
        original_ltv_pct: ltv,
        target_80_balance_usd: t80,
        target_78_balance_usd: t78,
        months_to_80: months_until(input.original_loan_usd, i, pmt, t80),
        months_to_78: months_until(input.original_loan_usd, i, pmt, t78),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-2
    }

    fn base() -> PmiInput {
        PmiInput {
            original_home_value_usd: 400_000.0,
            original_loan_usd: 360_000.0,
            annual_rate_pct: 6.0,
            loan_term_months: 360.0,
        }
    }

    #[test]
    fn payment_and_ltv() {
        let r = analyze(&base());
        assert!(close(r.monthly_payment_usd, 2158.381891));
        assert!(close(r.original_ltv_pct, 90.0));
    }

    #[test]
    fn thresholds() {
        let r = analyze(&base());
        assert!(close(r.target_80_balance_usd, 320_000.0));
        assert!(close(r.target_78_balance_usd, 312_000.0));
    }

    #[test]
    fn months_to_eighty() {
        // Balance hits 320k at month 89 (88.91 → ceil).
        assert!(close(analyze(&base()).months_to_80.unwrap(), 89.0));
    }

    #[test]
    fn months_to_seventy_eight() {
        assert!(close(analyze(&base()).months_to_78.unwrap(), 103.0));
    }

    #[test]
    fn auto_termination_after_request() {
        let r = analyze(&base());
        // 78% is reached later than 80%.
        assert!(r.months_to_78.unwrap() > r.months_to_80.unwrap());
    }

    #[test]
    fn no_pmi_when_under_80_at_origination() {
        // 20% down → 80% LTV → at the request threshold immediately.
        let r = analyze(&PmiInput {
            original_home_value_usd: 400_000.0,
            original_loan_usd: 320_000.0,
            annual_rate_pct: 6.0,
            loan_term_months: 360.0,
        });
        assert!(close(r.months_to_80.unwrap(), 0.0));
    }

    #[test]
    fn zero_rate_linear_paydown() {
        // 360k loan, 0%, 360 months → $1000/mo principal. To 320k: 40k/1000 = 40.
        let r = analyze(&PmiInput {
            original_home_value_usd: 400_000.0,
            original_loan_usd: 360_000.0,
            annual_rate_pct: 0.0,
            loan_term_months: 360.0,
        });
        assert!(close(r.months_to_80.unwrap(), 40.0));
    }

    #[test]
    fn lower_down_payment_delays_removal() {
        let small_down = analyze(&PmiInput {
            original_home_value_usd: 400_000.0,
            original_loan_usd: 380_000.0,
            annual_rate_pct: 6.0,
            loan_term_months: 360.0,
        });
        assert!(small_down.months_to_80.unwrap() > analyze(&base()).months_to_80.unwrap());
    }
}
